### ISSUE TITLE
Revamping/Streamling Console Commands

### ISSUE THREAD

**Issue by cloudRoutine** [fsprojects/Forge#45]
As it stands I think the current commands are a bit clunky, but figuring out how to add existing and create new files/projects/solutions in a unified and straightforward manner is not trivial.

I think it'd be useful to shift to a verb-noun format removing the `file` and `reference` commands in favor of something like 

`add --file module1.fs --project project1 [--link] [--folder|--directory] [--build-action]` 
`add --file module2.fs --solution testsolution [--folder|--directory] [--build-action]`

`--link` is for the option of adding the file as a link
If both `--project` and `--solution` are specified the command should fail
`[--folder|--directory] ` is an optional arg to add it into a folder in the project or a solution folder
If there's only one project available and neither project or solution is specified should we assume that's what it should be added to or should we always prompt with a list of available options?

`new --file module3.fs --project project1 [--template] [--folder|--directory]` 
`new --file module3.fs --project project1 [--template] [--folder|--directory]`
`new --project project2.fsproj --solution testsolution [--template] [--folder|--directory]`

For these kinds of commands `--file` & `--project`, `--project` & `--solution`, and `--file` & `--solution` could be paired
Once item templates are supported `--template` selects from existing file templates

Should these kinds of commands create the directory if it doesn't exist?

`add --reference assembly1 --project project1.fsproj` 

However this would be an issue when trying to add a reference to another project in the solution so maybe `--target`,`--into` or some other catchall name for the destination would be better for a more consistent and clear command set.

There's a lot more to discuss beyond this, but I want to get the ball rolling.

Thoughts? Comments? Concerns?

**Comment by Krzysztof-Cieslak** [comment]
I actually like `something:something syntax` so would prefer `new:project` (maybe as default just `new` ?) `new:file`, `add:file`, `add:reference` and then additional options.

**Comment by cloudRoutine** [comment]
I think that's good, it fixes the double project problem

**Comment by Krzysztof-Cieslak** [comment]
So we would end up with something like that:

```
new:project [--name <string>] [--directory <string>] [--template <string>] [--no-paket] [--no-fake]
new:file [--name <string>] [--template <string>] [--project <string>]  [--solution <string>]  [--link] [--build-action <string>]
add:file [--name <string>]  [--project <string>]  [--solution <string>]  [--link] [--build-action <string>]
add:reference [--name <string>]  [--project <string>]
remove:file  [--name <string>]  [--project <string>]
remove:reference  [--name <string>]  [--project <string>]
rename:file [--name <string>] [--new <string>]
rename:project [--name <string>] [--new <string>]
list:file [--project <string>]
list:reference [--project <string>]
update:paket
update:fake
paket
fake
refresh
exit
```

Important thing to be decided - how to handle "internal" directories / paths for commands involving files. Initial assumption (and I would like to stick with it) was to always run ~Fix~ Forge from solution/root folder. In such case something like file remove would look `forge remove:file -n src\project\file.fs -p src\project\proj.fsproj`. Or would it better to add "common" `--directory` parameter which would just change relative paths for every command - `forge remove:file -d src\project -n file.fs -p proj.fsproj`

**Comment by Tarmil** [comment]
Yes I am very much in favor of always running forge from the root folder. Always passing `-p src/foo/foo.fsproj` would be tedious, but in 99% of cases we can easily guess it:
* For `*:file` commands, find a `.fsproj` file in the same directory as the given file, if not found search up the tree until reaching the working directory, and if we reach the working directory without finding any then we fail. For example, if you're in the root of the Forge solution and run `new:file -n src/Forge.Core/Utils/Utils.fs` then we create that file and add it to `src/Forge.Core/Forge.Core.fsproj` as `Utils/Utils.fs`.
* For all commands, if `-p` receives a directory then we find a project file in that directory or fail if there are none or several.

**Comment by reidev275** [comment]
This looks like a very good suggestion to me because it brings solutions into the discussion which have so far been underrepresented. 

@cloudRoutine Verb -> Noun vs Noun -> Verb seems more of a stylistic choice but I really have no strong feelings either way.  My initial thought was more like git where you could do `file --help` and get the options possible for the file command.  I think the difference is akin to whether you like `fn x` or `x |> fn`.  Starting with the Verb feels typeclassy to me.  

@Krzysztof-Cieslak having an optional `--directory` `-d` parameter is a really cool idea that probably wouldn't be all that difficult to add and could make it easier for deeply nested solutions.

Agreed with @Tarmil on the guessing of the project folder.  If we can't find a project based upon the strategies then we could either prompt for the project or refuse the command and force the user to reissue the command with the `-p` option.

**Comment by Krzysztof-Cieslak** [comment]
Problem: Argue right now doesn't support `something:something`, `:` is not valid character

**Comment by Tarmil** [comment]
Actually there is no reason we should necessarily use `:`. How about `-`? It's a bit easier to type because you don't need to press shift.

**Comment by Krzysztof-Cieslak** [comment]
Maybe just using space? 

```
new project [--name <string>] [--directory <string>] [--template <string>] [--no-paket] [--no-fake]
new file [--name <string>] [--template <string>] [--project <string>]  [--solution <string>]  [--link] [--build-action <string>]
add file [--name <string>]  [--project <string>]  [--solution <string>]  [--link] [--build-action <string>]
add reference [--name <string>]  [--project <string>]
remove file  [--name <string>]  [--project <string>]
remove reference  [--name <string>]  [--project <string>]
rename file [--name <string>] [--new <string>]
rename project [--name <string>] [--new <string>]
list:file [--project <string>]
list:reference [--project <string>]
update paket
update fake
paket
fake
refresh
exit
```

**Comment by Tarmil** [comment]
Yeah that would work too.

**Comment by Krzysztof-Cieslak** [comment]
WIP PR for that is #49

**Comment by smoothdeveloper** [comment]
I favor verb -> noun, seems more in line with git for exemple.

**Comment by reidev275** [comment]
funny @smoothdeveloper it was git I was thinking of too, but apparently a different part :)
e.g. `git remote add`

**Comment by Krzysztof-Cieslak** [comment]
Implemented in #49
